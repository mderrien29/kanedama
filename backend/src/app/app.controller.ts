import { Controller, Get } from '@nestjs/common';

import { AppService } from './app.service';
import { AnswerDto } from 'src/common/dtos/answer.dto';

@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get()
  getHello(): string {
    return this.appService.getHello();
  }

  @Get('answer')
  answer(): Promise<AnswerDto> {
    return this.appService.getAnswer();
  }
}
