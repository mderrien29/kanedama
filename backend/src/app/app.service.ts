import { Injectable } from '@nestjs/common';
import { AccountsService } from './accounts/accounts.service';

@Injectable()
export class AppService {
  constructor(private readonly accountsService: AccountsService) {}

  getHello(): string {
    return 'Hello World!';
  }

  public async getAnswer(): Promise<string> {
    // TODO getApiData ?
    const accounts = await this.accountsService.getUserAccounts();
    const currentAccountValue =
      this.accountsService.sumAccountsCurrentAmount(accounts);
    const transactions = await this.accountsService.getUserFullHistory(
      accounts,
    );
    const transactionsOrderedByDate =
      this.accountsService.sortTransactionsByDate(transactions);

    // TODO calculateMetrics ?
    const amountOverTime =
      this.accountsService.calculateAccountAmountOverTimeFromCurrent(
        transactionsOrderedByDate,
        currentAccountValue,
      );
    const [min, max] =
      this.accountsService.calculateRoundedMinMax(amountOverTime);
    const hasRecentActivity = this.accountsService.hasRecentActivity(
      transactionsOrderedByDate,
      new Date(new Date().setFullYear(new Date().getFullYear() - 3)), // TODO hide this
    );
    const averageRecentIncome =
      this.accountsService.calculateRoundedAverageIncome(transactions, 6);

    // TODO formatResults. maybe round there ?
    return JSON.stringify(
      this.formatResults(min, max, averageRecentIncome, hasRecentActivity),
    );
  }

  // TODO any
  private formatResults(
    min: number,
    max: number,
    averageIncome: number,
    activeLast3Years: boolean,
  ): any {
    return {
      '6_month_average_income': averageIncome,
      '3_years_activity': activeLast3Years,
      max_balance: max,
      min_balance: min,
    };
  }
}
