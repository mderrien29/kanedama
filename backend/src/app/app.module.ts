import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';

import { AccountsModule } from './accounts/accounts.module';
import { AppController } from './app.controller';
import { AppService } from './app.service';

@Module({
  imports: [
    AccountsModule,
    ConfigModule.forRoot({ isGlobal: true, envFilePath: '.env.dev' }),
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
