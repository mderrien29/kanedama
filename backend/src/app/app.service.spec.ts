import { Test, TestingModule } from '@nestjs/testing';

import { AppService } from './app.service';
import * as userAccounts from '../../test/resources/user-accounts.json';
import * as userTransactions from '../../test/resources/user-transactions.json';
import * as userMetrics from '../../test/resources/user-metrics.json';

describe('AppService', () => {
  let appService: AppService;
  const accountsServiceMock = {
    getUserAccounts: jest.fn(() => userAccounts),
    getAllTransactions: jest.fn(() => userTransactions),
  };

  const metricsServiceMock = {
    getUserMetrics: jest.fn(() => userMetrics),
  };

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      providers: [
        AppService,
        { provide: 'AccountsService', useFactory: () => accountsServiceMock },
        { provide: 'MetricsService', useFactory: () => metricsServiceMock },
      ],
    }).compile();

    appService = app.get<AppService>(AppService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('getAnswer', () => {
    it('should call AppService with the expected parameters', async () => {
      await appService.getAnswer();

      expect(accountsServiceMock.getUserAccounts).toHaveBeenCalled();
      expect(accountsServiceMock.getAllTransactions).toHaveBeenCalled();
      expect(
        accountsServiceMock.getAllTransactions.mock.calls,
      ).toMatchSnapshot();
    });

    it('should call MetricsService with the expected parameters', async () => {
      await appService.getAnswer();
      expect(metricsServiceMock.getUserMetrics).toHaveBeenCalled();
      expect(metricsServiceMock.getUserMetrics.mock.calls).toMatchSnapshot();
    });

    it('should return the result from MetricsService', async () => {
      const answer = await appService.getAnswer();
      expect(answer).toMatchSnapshot();
    });

    it('should throw human readable error if API fails', async () => {
      accountsServiceMock.getAllTransactions.mockImplementationOnce(() => {
        throw new Error('mock');
      });

      // rejects does not work, related issue : https://github.com/facebook/jest/issues/4946
      appService.getAnswer().catch((e) => expect(e.response).toMatchSnapshot());
    });

    it('should throw human readable error if calculations fails', async () => {
      metricsServiceMock.getUserMetrics.mockImplementationOnce(() => {
        throw new Error('mock');
      });

      appService.getAnswer().catch((e) => expect(e.response).toMatchSnapshot());
    });
  });
});
