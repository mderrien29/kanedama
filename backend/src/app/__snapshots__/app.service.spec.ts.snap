// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppService getAnswer should call AppService with the expected parameters 1`] = `
Array [
  Array [
    Array [
      Object {
        "account_id": "id",
        "account_number": "1234",
        "account_type": "ACCOUNT",
        "available": 10000,
        "currency": "USD",
        "current": 10000,
        "iban": "iban",
        "sort_code": "sort",
        "swift_bic": "bic",
        "update_timestamp": "2021-08-30",
      },
    ],
  ],
]
`;

exports[`AppService getAnswer should call MetricsService with the expected parameters 1`] = `
Array [
  Array [
    Array [
      Object {
        "account_id": "id",
        "account_number": "1234",
        "account_type": "ACCOUNT",
        "available": 10000,
        "currency": "USD",
        "current": 10000,
        "iban": "iban",
        "sort_code": "sort",
        "swift_bic": "bic",
        "update_timestamp": "2021-08-30",
      },
    ],
    Array [
      Object {
        "amount": 1000,
        "currency": "GBP",
        "status": "status",
        "timestamp": "2019-08-30",
        "transaction_category": "category",
        "transaction_type": "type",
      },
      Object {
        "amount": 10,
        "currency": "GBP",
        "status": "status",
        "timestamp": "2020-08-30",
        "transaction_category": "category",
        "transaction_type": "type",
      },
      Object {
        "amount": -20,
        "currency": "GBP",
        "status": "status",
        "timestamp": "2020-08-30",
        "transaction_category": "category",
        "transaction_type": "type",
      },
      Object {
        "amount": 100,
        "currency": "GBP",
        "status": "status",
        "timestamp": "2020-08-31",
        "transaction_category": "category",
        "transaction_type": "type",
      },
      Object {
        "amount": 5,
        "currency": "GBP",
        "status": "status",
        "timestamp": "2020-10-20",
        "transaction_category": "category",
        "transaction_type": "type",
      },
      Object {
        "amount": -100,
        "currency": "GBP",
        "status": "status",
        "timestamp": "2020-11-01",
        "transaction_category": "category",
        "transaction_type": "type",
      },
    ],
  ],
]
`;

exports[`AppService getAnswer should return the result from MetricsService 1`] = `
Object {
  "3_years_activity": true,
  "6_month_average_income": 407,
  "max_balance": 19540,
  "min_balance": -4285,
}
`;

exports[`AppService getAnswer should throw human readable error if API fails 1`] = `
Object {
  "error": "Application failed to retrieve data from API, due to Error: mock",
  "status": 500,
}
`;

exports[`AppService getAnswer should throw human readable error if calculations fails 1`] = `
Object {
  "error": "Application failed to calculate metrics, due to Error: mock",
  "status": 500,
}
`;
